asynciterator.ts

```diff
@@ -1252,6 +1254,135 @@ function destinationFillBuffer<S>(this: InternalSource<S>) {
+      this.emit('error', err);
+    };
+    const onSourceReadable = () => {
+      if (this.readable) {
+        // TODO: I'm not completely sure as to why this is needed but without
+        //       the following line, some use cases relying on flow mode (i.e.
+        //       consuming items via `on('data', (data) => {})`) do not work.
+        //       It looks like the debouncing that happens in `set readable()`
+        //       in `AsyncIterator` prevents the event from firing as `this`
+        //       is already readable.
+        this.emit('readable');
+      }
+      else {
+        this.readable = true;
+      }
+    };
+    _source.on('end', onSourceEnd);
+    _source.on('error', onSourceError);
+    _source.on('readable', onSourceReadable);
+    if (_source.done)
+      onSourceEnd();
+    else if (_source.readable)
+      onSourceReadable();
+  }
+
+  destroy(cause?: Error): void {
+    this._source.destroy(cause);
+    super.destroy(cause);
+  }
+
+  public close() {
+    if (this._destroySource)
+      this._source.destroy();
+    super.close();
+  }
+}
+
+interface Transform {
+  fn: Function,
+  next?: Transform
+}
+
+export class SyncTransformIterator<T, D = T> extends SynchronousTransformIterator<T, D> {
+  private _funcs?: Function[];
+
+  constructor(private source: AsyncIterator<T>, private transforms: Transform, private upstream: AsyncIterator<any> = source, options?: { destroySource?: boolean }) {
+    // Subscribe the iterator directly upstream rather than the original source to avoid over-subscribing
+    // listeners to the original source
+    super(upstream, options);
+  }
+
+  get funcs() {
+    if (!this._funcs) {
+      this._funcs = [];
+      // eslint-disable-next-line prefer-destructuring
+      let transforms: Transform | undefined = this.transforms;
+      do
+        this._funcs.push(transforms.fn);
+      // eslint-disable-next-line no-cond-assign
+      while (transforms = transforms.next);
+    }
+    return this._funcs;
+  }
+
+  read(): D | null {
+    const { source, funcs } = this;
+    let item;
+    outer: while ((item = source.read()) !== null) {
+      // Do not use a for-of loop here, it slows down transformations
+      // by approximately a factor of 2.
+      for (let index = funcs.length - 1; index >= 0; index -= 1) {
+        if ((item = funcs[index](item)) === null)
+          continue outer;
+      }
+      return item;
+    }
+    return null;
+  }
+
+  map<K>(map: (item: D) => K | null, self?: any): AsyncIterator<K> {
+    return new SyncTransformIterator<T, K>(this.source, { fn: self ? map.bind(self) : map, next: this.transforms }, this);
+  }
+
+  destroy(cause?: Error): void {
+    this.upstream.destroy(cause);
+    super.destroy(cause);
+  }
+
+  public close() {
+    if (this.options.destroySource)
+      this.upstream.destroy();
+    super.close();
+  }
+}
+
+export class LimitingIterator<T> extends SynchronousTransformIterator<T> {
```

Comment:
Why not just directly from `AsyncIterator`, or `TransformIterator`?
All we need is the `source` mechanics.

Reply:
The source mechanics from the `TransformIterator` are a bit heavyweight esp since it extends from the `BufferedIterator`.